File: Code/C++/trappingRainwater.cpp
/* Aim: Find total units of rainwater that can be stored for the given array of heights
Approach Used : 1. Find maximum height of bar from the left end upto an index i in the array left_max.
2. Find maximum height of bar from the right end upto an index i in the array right_max.
3. Iterate over the array and update ans: Add min(left_max[i],right_max[i]) - height[i] to ans */
#include <bits/stdc++.h>
using namespace std;
int trappingRainwater(vector<int>& height) {
    int res = 0,left,right;
    int n = height.size();
    for(int i =1; i<n-1; i++){
        left = height[i];
        for(int j =0; j<i; j++){
            left =max(height[j],left);
        }        
        right = height[i];
        for(int j =i+1; j<n; j++){
            right =max(height[j],right);
        }        
        res+= min(left,right) - height[i];      
    }
    return res;
}
int main() {
    int n;
    cout<<"Enter Size of Array : ";
    cin>>n;
    vector<int> arr(n,0);
    for(int i = 0; i < n; i++) 
        cin>>arr[i];
    cout << trappingRainwater(arr);
    return 0;
}
/* Time complexity: O(n)
Space complexity: O(n)
Test Case 1:
Input - 12
        0 1 0 2 1 0 1 3 2 1 2 1
Output - 6 
Test Case 2:
Input - 6
        420325
Output - 9 */

File: Code/Python/trapping_rainwater.py
"""
Problem Description:
Given an integer array A of non-negative integers representing an elevation map
where the width of each bar is 1, compute maximum water it is able to trap after raining.
"""

# Function to calculate the maximum water that can be trapped .
def max_water(arr, n):

    temp = 0
    # iterate the array
    for i in range(1, n - 1):

        # Find the element with maximum height on right of element i
        max_right = arr[i]

        for j in range(i+1, n):
            max_right = max(max_right, arr[j])

        # Find the element with maximum height on the left of element i
        max_left = arr[i]
        for j in range(i):
            max_left = max(max_left, arr[j])

        # Updating the temp value
        temp = temp + (min(max_right, max_left) - arr[i])

    return temp


if __name__ == "__main__":
    arr = []
    print("Enter number of elements")
    # number of elements as input
    n = int(input())
    
    print("Enter the elements")
    # iterating till the range
    for i in range(0, n):
        p = int(input())
        arr.append(p)
    print(max_water(arr, n))

"""
Test case 1:
    input:
    n=5
    arr=[2,0,2,3,4]
    output:
    2

Test case 2:
    input:
    n=7
    arr=[5,1,2,3,2,6,1]
    output:
    12

Time Complexity: O(n)

Space Complexity: O(1)
"""
